bgbb/sql/sql_utils.py

```diff
@@ -0,0 +1,219 @@
+    """
+    iter of ints to SQL string version for main_summary
+    >>> to_samp_ids([0, 1, 2])
+    "'0', '1', '2'"
+    """
+    mn = min(samp_ids)
+    mx = max(samp_ids)
+    if mn == 1:
+        print('Warning: 0 is minimum sample_id allowed, not 1')
+    elif mn < 0:
+        raise ValueError('sample_id >= "0" in main summary')
+    if mx > 99:
+        raise ValueError('sample_id <= "99" in main summary')
+    return to_sql_list(map(str, samp_ids))
+    # (', '.join("'{}'".format(i) for i in samp_ids))
+
+
+def to_sql_list(xs):
+    """stringify lists for SQL queries
+    >>> to_sql_list([1, 2, 3]) == '1, 2, 3'
+    """
+    def to_sql_literal(x):
+        if isinstance(x, str):
+            return "'{}'".format(x)
+        return str(x)
+    res = ', '.join(map(to_sql_literal, xs))
+    return res
+
+
+def test_to_sql_list():
+    assert to_sql_list(['GB']) == "'GB'"
+    assert to_sql_list(['GB', 'US', 'IN']) == "'GB', 'US', 'IN'"
+    assert to_sql_list([1, 2, 3]) == '1, 2, 3'
+
+
+def insert_country(q, insert_before="{sample_comment}",
+                   countries: List[str]=["GB"]):
+    "Insert country restriction into SQL string for testing"
+    i = q.find(insert_before)
+    to_insert = "AND country IN ({})\n      ".format(to_sql_list(countries))
+    return q[:i] + to_insert + q[i:]
+
+
+def mk_time_params(HO_WIN=14, MODEL_WIN=90, ho_start='2018-08-01'):
+    """
+    Return container whose attributes are holdout and model input
+    date ranges, specified by a training window `MODEL_WIN`,
+    holdout evaluation window `HO_WIN` and holdout start date `ho_start`
+    (day after last day in model window).
+    """
+    def r():
+        pass
+
+    r.ho_start_datet = pd.to_datetime(ho_start)
+    r.ho_start_date = r.ho_start_datet.date()
+    r.ho_last_date = r.ho_start_date + dt.timedelta(days=HO_WIN - 1)
+    r.model_start_date = r.ho_start_date - dt.timedelta(days=MODEL_WIN)
+
+    # Str format
+    mod_ho_ev = [r.model_start_date, r.ho_start_date, r.ho_last_date]
+    # mod_ho_ev_str
+    (r.model_start_date_str, r.ho_start_date_str, r.ho_last_date_str
+     ) = map(to_s3_fmt, mod_ho_ev)
+
+    test_ho_range(r.ho_start_date, r.ho_last_date, HO_WIN)
+    test_model_range(model_start_date=r.model_start_date,
+                     ho_start_date=r.ho_start_date,
+                     MODEL_WIN=MODEL_WIN)
+
+    # r.__dict__.update(locals())
+    return r
+
+
+base_query = '''
+with cid_day as (
+    SELECT
+        C.client_id
+        , C.submission_date_s3
+        , from_unixtime(unix_timestamp(C.submission_date_s3, 'yyyyMMdd'),
+                        'yyyy-MM-dd') AS sub_date
+    FROM clients_daily C
+    WHERE
+        app_name = 'Firefox'
+        AND channel = 'release'
+      {sample_comment}  AND sample_id in ({sample_ids})
+)
+
+-- clients_daily aggregates from window *before*
+-- the holdout date
+, cid_model as (
+    SELECT
+        C.client_id
+        , MIN(C.sub_date) Min_day
```

Comment:
There's inconsistent usage of aliasing in here, I would use `AS` for consistency.

Reply:
👍